samples/JobsAgent/JobsAgentOperations.cpp

```diff
@@ -0,0 +1,693 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/**
+ * @file JobsAgentOperations.cpp
+ *
+ */
+
+#include <chrono>
+#include <cstring>
+#include <sys/utsname.h>
+#include <fstream>
+#include <signal.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "util/logging/Logging.hpp"
+#include "util/logging/LogMacros.hpp"
+#include "util/logging/ConsoleLogSystem.hpp"
+
+#include "ConfigCommon.hpp"
+#include "JobsAgent.hpp"
+
+#ifdef UNIT_TESTS
+#define IF_UNIT_TESTS(stmt) stmt
+#define IF_NOT_UNIT_TESTS(stmt)
+#else
+#define IF_UNIT_TESTS(stmt)
+#define IF_NOT_UNIT_TESTS(stmt) stmt
+#endif
+
+/*
+jobs agent error values:
+
+ERR_DOWNLOAD_FAILED
+ERR_FILE_COPY_FAILED
+ERR_UNNAMED_PACKAGE
+ERR_INVALID_PACKAGE_NAME
+ERR_SYSTEM_CALL_FAILED
+ERR_UNEXPECTED_PACKAGE_EXIT
+ERR_UNABLE_TO_START_PACKAGE
+ERR_UNABLE_TO_STOP_PACKAGE
+ERR_UNSUPPORTED_CHECKSUM_ALGORITHM
+ERR_CHECKSUM_FAILED
+ERR_UNEXPECTED
+*/
+
+namespace awsiotsdk {
+    namespace samples {
+        const util::String JobsAgent::INSTALLED_PACKAGES_FILENAME = "installedPackages.json";
+
+        void JobsAgent::ShowJobsError(util::String operation, ResponseCode rc) {
+            if (ResponseCode::SUCCESS != rc) {
+                util::String fullMessage = "Error in %s operation. %s";
+                AWS_LOG_ERROR(LOG_TAG_JOBS_AGENT, fullMessage.c_str(), 
+                              operation.c_str(), ResponseHelper::ToString(rc).c_str());
+            }            
+        }
+
+        util::String JobsAgent::GetShutdownSystemCommand(bool dryRun, bool reboot) {
+            IF_UNIT_TESTS(if (!use_valid_system_command_) return "invalid command";)
+
+            util::String result = "sudo /sbin/shutdown ";
```

Comment:
what about windows?

Reply:
if you need to call sudo, isn't this going to need permissions. shouldn't it at least be gksudo?

Also, the assumption is the caller has sudo permissions, why not just set the sudo bit on the program?

Reply:
sudo permissions can be set for individual commands, so I believe this is a reasonable approach. Also added version for Windows.